cura/Settings/MachineManager.py

```diff
@@ -517,6 +517,12 @@ def activeMachineId(self) -> str:
             return self._global_container_stack.getId()
         return ""
 
+    @pyqtProperty(str, notify = globalContainerChanged)
+    def activeMachineFirmwareVersion(self) -> str:
+        if not self._printer_output_devices[0]:
+            return ""
+        return self._printer_output_devices[0].firmwareVersion
+
```

Comment:
`active_machine = CuraApplication.getInstance().global_container_stack` and `active_machine = CuraApplication.getInstance().getMachineManager().activeMachine` are both used all over our codebase. Unsure which one is correct. I opted for the machine manager approach. However I'm also unclear about if we should be using the global stack for firmware. Checking if the machine has a network connection uses `self._printer_output_devices` but for almost everything else we use `self._global_container_stack.getWhatever()`. So I could write some some methods for the global stack to handle firmware but it feels a bit pointless right now because there's no prevailing way to handle the "current machine".

Reply:
There is no `active_machine = CuraApplication.getInstance().global_container_stack` There is a `_global_container_stack`, but that is protected (and should not be used outside of `Application` and it's children.

Reply:
Sorry. Typo on my part. It's unclear which is the right way to get the machine:

- `active_machine = CuraApplication.getInstance().getGlobalContainerStack`, which returns the global container stack, or...

-  `active_machine = CuraApplication.getInstance().getMachineManager().activeMachine`, which also returns the global container stack.